src/server/pps/server/api_server.go

```diff
@@ -2370,74 +2318,52 @@ func (a *apiServer) listPipeline(ctx context.Context, request *pps.ListPipelineR
 		return true
 	}
 
-	// fetch details and filter via a task chain so that order is preserved
-	chain := chunk.NewTaskChain(ctx)
-
-	// spin up goroutines to get the PFS info
-	limiter := limit.New(5)
 	for i := range infos {
-		i := i
-		limiter.Acquire()
-		if err := chain.CreateTask(func(ctx context.Context, serial func(func() error) error) error {
-			defer limiter.Release()
-			if request.Details {
-				if err := ppsutil.GetPipelineDetails(a.env.GetPachClient(ctx), infos[i]); err != nil {
-					return err
-				}
+		if filterPipeline(infos[i]) {
+			if err := f(infos[i]); err != nil {
+				return err
 			}
-			return serial(func() error {
-				if filterPipeline(infos[i]) {
-					return f(infos[i])
-				}
-				return nil
-			})
-		}); err != nil {
-			return err
 		}
 	}
-	return chain.Wait()
+	return nil
 }
 
 // listPipelineInfo enumerates all PPS pipelines in the database, filters them
 // based on 'request', and then calls 'f' on each value
 func (a *apiServer) listPipelineInfo(ctx context.Context,
 	pipeline *pps.Pipeline, history int64, f func(*pps.PipelineInfo) error) error {
 	p := &pps.PipelineInfo{}
-	forEachPipeline := func() error {
+	versionMap := make(map[string]uint64)
+	checkPipelineVersion := func() error {
 		// Erase any AuthToken - this shouldn't be returned to anyone (the workers
 		// won't use this function to get their auth token)
 		p.AuthToken = ""
 		// TODO: this is kind of silly - callers should just make a version range for each pipeline?
-		lastVersionToSend := int64(p.Version) - history
-		if history == -1 || lastVersionToSend < 1 {
-			lastVersionToSend = 1
-		}
-		for int64(p.Version) >= lastVersionToSend {
-			if err := f(p); err != nil {
-				return err
+		if current, ok := versionMap[p.Pipeline.Name]; ok {
+			if p.Version < current {
+				// don't send, exit early
+				return nil
 			}
-			p.Version -= 1
-		}
-		return nil
-	}
-	if pipeline == nil {
-		if err := a.pipelines.ReadOnly(ctx).List(p, col.DefaultOptions(), func(string) error {
-			return forEachPipeline()
-		}); err != nil {
-			return err
-		}
-	} else {
-		if err := a.pipelines.ReadOnly(ctx).Get(pipeline.Name, p); err != nil {
-			if col.IsErrNotFound(err) {
-				return errors.Errorf("pipeline \"%s\" not found", pipeline.Name)
+		} else {
+			// we haven't seen this pipeline yet, rely on sort order and assume this is latest
+			var lastVersionToSend uint64
+			if history < 0 || uint64(history) > p.Version {
```

Comment:
I think the second predicate here should be ">=", right?

Reply:
Ah probably true, although I think it works as is because both 1 and 0 will send everything. In any case I should add a test for this to make sure.

Reply:
Ah probably true, although I think it works as is because both 1 and 0 will send everything. In any case I should add a test for this to make sure.